File: routes/inventory.py
from flask import Blueprint, render_template, jsonify, request
from app import db
from models import Product, StockMovement
import logging

# Create a blueprint for inventory management
inventory_bp = Blueprint('inventory', __name__, url_prefix='/inventory')

@inventory_bp.route('/')
def inventory_page():
    """Display the inventory management page"""
    return render_template('inventory.html')

@inventory_bp.route('/products', methods=['GET'])
def get_products():
    """API endpoint to get all products"""
    try:
        products = Product.query.all()
        return jsonify({
            'success': True,
            'products': [product.to_dict() for product in products]
        })
    except Exception as e:
        logging.error(f"Error fetching products: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@inventory_bp.route('/products/<int:product_id>', methods=['GET'])
def get_product(product_id):
    """API endpoint to get a specific product"""
    try:
        product = Product.query.get_or_404(product_id)
        return jsonify({
            'success': True,
            'product': product.to_dict()
        })
    except Exception as e:
        logging.error(f"Error fetching product {product_id}: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@inventory_bp.route('/products', methods=['POST'])
def add_product():
    """API endpoint to add a new product"""
    try:
        data = request.json
        product = Product(
            barcode=data.get('barcode'),
            name=data.get('name'),
            description=data.get('description', ''),
            price=float(data.get('price')),
            cost_price=float(data.get('cost_price')),
            category=data.get('category'),
            stock_quantity=int(data.get('stock_quantity', 0))
        )
        db.session.add(product)
        db.session.commit()
        
        # Create stock movement for initial stock
        if product.stock_quantity > 0:
            stock_movement = StockMovement(
                product_id=product.id,
                quantity=product.stock_quantity,
                movement_type='in',
                reference='Initial Stock',
                notes='Initial stock upon product creation'
            )
            db.session.add(stock_movement)
            db.session.commit()
            
        return jsonify({
            'success': True,
            'product': product.to_dict()
        }), 201
    except Exception as e:
        db.session.rollback()
        logging.error(f"Error adding product: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@inventory_bp.route('/products/<int:product_id>', methods=['PUT'])
def update_product(product_id):
    """API endpoint to update a product"""
    try:
        product = Product.query.get_or_404(product_id)
        data = request.json
        
        # Track stock changes for movement record
        old_stock = product.stock_quantity
        
        # Update product fields if present in the request
        if 'barcode' in data:
            product.barcode = data['barcode']
        if 'name' in data:
            product.name = data['name']
        if 'description' in data:
            product.description = data['description']
        if 'price' in data:
            product.price = float(data['price'])
        if 'cost_price' in data:
            product.cost_price = float(data['cost_price'])
        if 'category' in data:
            product.category = data['category']
        if 'stock_quantity' in data:
            product.stock_quantity = int(data['stock_quantity'])
        
        db.session.commit()
        
        # Create stock movement if quantity changed
        new_stock = product.stock_quantity
        if new_stock != old_stock:
            movement_type = 'in' if new_stock > old_stock else 'out'
            quantity = abs(new_stock - old_stock)
            stock_movement = StockMovement(
                product_id=product.id,
                quantity=quantity if movement_type == 'in' else -quantity,
                movement_type=movement_type,
                reference='Stock Adjustment',
                notes=f'Stock adjusted from {old_stock} to {new_stock}'
            )
            db.session.add(stock_movement)
            db.session.commit()
            
        return jsonify({
            'success': True,
            'product': product.to_dict()
        })
    except Exception as e:
        db.session.rollback()
        logging.error(f"Error updating product {product_id}: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@inventory_bp.route('/products/<int:product_id>', methods=['DELETE'])
def delete_product(product_id):
    """API endpoint to delete a product"""
    try:
        product = Product.query.get_or_404(product_id)
        db.session.delete(product)
        db.session.commit()
        return jsonify({
            'success': True,
            'message': f'Product {product_id} deleted successfully'
        })
    except Exception as e:
        db.session.rollback()
        logging.error(f"Error deleting product {product_id}: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@inventory_bp.route('/stock-movements', methods=['GET'])
def get_stock_movements():
    """API endpoint to get stock movement history"""
    try:
        movements = StockMovement.query.order_by(StockMovement.movement_date.desc()).all()
        return jsonify({
            'success': True,
            'movements': [movement.to_dict() for movement in movements]
        })
    except Exception as e:
        logging.error(f"Error fetching stock movements: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@inventory_bp.route('/stock-movements', methods=['POST'])
def add_stock_movement():
    """API endpoint to add a stock movement"""
    try:
        data = request.json
        product = Product.query.get_or_404(data.get('product_id'))
        quantity = int(data.get('quantity', 0))
        movement_type = data.get('movement_type')
        
        # Create the stock movement
        stock_movement = StockMovement(
            product_id=product.id,
            quantity=quantity if movement_type == 'in' else -quantity,
            movement_type=movement_type,
            reference=data.get('reference', ''),
            notes=data.get('notes', '')
        )
        db.session.add(stock_movement)
        
        # Update product stock quantity
        if movement_type == 'in':
            product.stock_quantity += quantity
        elif movement_type == 'out':
            product.stock_quantity -= quantity
        
        db.session.commit()
        
        return jsonify({
            'success': True,
            'movement': stock_movement.to_dict(),
            'updated_product': product.to_dict()
        }), 201
    except Exception as e:
        db.session.rollback()
        logging.error(f"Error adding stock movement: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@inventory_bp.route('/categories', methods=['GET'])
def get_categories():
    """API endpoint to get all product categories"""
    try:
        # Get unique categories from products
        categories = db.session.query(Product.category).distinct().all()
        category_list = [category[0] for category in categories]
        
        return jsonify({
            'success': True,
            'categories': category_list
        })
    except Exception as e:
        logging.error(f"Error fetching categories: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

File: static/js/inventory.js
/**
 * Inventory Management JavaScript
 * Handles product and stock management functionality
 */

// Inventory state
const InventoryState = {
    products: [],
    categories: [],
    stockMovements: [],
    selectedProduct: null,
    currentView: 'products', // 'products' or 'stock'
};

// DOM Elements
document.addEventListener('DOMContentLoaded', function() {
    if (document.getElementById('inventoryPage')) {
        initializeInventory();
    }
});

/**
 * Initialize the inventory page
 */
async function initializeInventory() {
    try {
        // Load products and categories
        await Promise.all([
            loadProducts(),
            loadCategories(),
            loadStockMovements()
        ]);
        
        // Set up event listeners
        setupInventoryEventListeners();
        
        // Show products by default
        showProductsView();
    } catch (error) {
        console.error('Error initializing inventory:', error);
        showToast('Error', 'Failed to initialize inventory', 'error');
    }
}

/**
 * Set up event listeners for inventory page
 */
function setupInventoryEventListeners() {
    // Tab navigation
    document.getElementById('productsTab').addEventListener('click', showProductsView);
    document.getElementById('stockTab').addEventListener('click', showStockView);
    
    // Product form
    const productForm = document.getElementById('productForm');
    if (productForm) {
        productForm.addEventListener('submit', handleProductFormSubmit);
    }
    
    // Stock movement form
    const stockForm = document.getElementById('stockMovementForm');
    if (stockForm) {
        stockForm.addEventListener('submit', handleStockMovementSubmit);
    }
    
    // New product button
    const newProductBtn = document.getElementById('newProductBtn');
    if (newProductBtn) {
        newProductBtn.addEventListener('click', () => {
            resetProductForm();
            const productModal = new bootstrap.Modal(document.getElementById('productModal'));
            productModal.show();
        });
    }
    
    // New stock movement button
    const newStockBtn = document.getElementById('newStockMovementBtn');
    if (newStockBtn) {
        newStockBtn.addEventListener('click', () => {
            resetStockMovementForm();
            const stockModal = new bootstrap.Modal(document.getElementById('stockMovementModal'));
            stockModal.show();
        });
    }
    
    // Search input
    const searchInput = document.getElementById('inventorySearch');
    if (searchInput) {
        searchInput.addEventListener('input', handleInventorySearch);
    }
    
    // Category filter
    const categoryFilter = document.getElementById('categoryFilter');
    if (categoryFilter) {
        categoryFilter.addEventListener('change', handleCategoryFilter);
    }
}

/**
 * Show the products view
 */
function showProductsView() {
    InventoryState.currentView = 'products';
    
    // Update active tab
    document.getElementById('productsTab').classList.add('active');
    document.getElementById('stockTab').classList.remove('active');
    
    // Show/hide relevant containers
    document.getElementById('productsContainer').style.display = 'block';
    document.getElementById('stockContainer').style.display = 'none';
    
    // Update buttons
    document.getElementById('newProductBtn').style.display = 'block';
    document.getElementById('newStockMovementBtn').style.display = 'none';
    
    // Refresh product list
    renderProductList();
}

/**
 * Show the stock view
 */
function showStockView() {
    InventoryState.currentView = 'stock';
    
    // Update active tab
    document.getElementById('productsTab').classList.remove('active');
    document.getElementById('stockTab').classList.add('active');
    
    // Show/hide relevant containers
    document.getElementById('productsContainer').style.display = 'none';
    document.getElementById('stockContainer').style.display = 'block';
    
    // Update buttons
    document.getElementById('newProductBtn').style.display = 'none';
    document.getElementById('newStockMovementBtn').style.display = 'block';
    
    // Refresh stock movement list
    renderStockMovements();
}

/**
 * Load products from API
 */
async function loadProducts() {
    try {
        const response = await apiCall('/inventory/products');
        InventoryState.products = response.products || [];
        renderProductList();
    } catch (error) {
        console.error('Error loading products:', error);
    }
}

/**
 * Load categories from API
 */
async function loadCategories() {
    try {
        const response = await apiCall('/inventory/categories');
        InventoryState.categories = response.categories || [];
        
        // Populate category dropdowns
        populateCategoryDropdowns();
    } catch (error) {
        console.error('Error loading categories:', error);
    }
}

/**
 * Load stock movements from API
 */
async function loadStockMovements() {
    try {
        const response = await apiCall('/inventory/stock-movements');
        InventoryState.stockMovements = response.movements || [];
        renderStockMovements();
    } catch (error) {
        console.error('Error loading stock movements:', error);
    }
}

/**
 * Populate category dropdowns
 */
function populateCategoryDropdowns() {
    const categoryFilter = document.getElementById('categoryFilter');
    const productCategory = document.getElementById('productCategory');
    
    if (categoryFilter) {
        // Clear current options except for the first one (All Categories)
        while (categoryFilter.options.length > 1) {
            categoryFilter.remove(1);
        }
        
        // Add categories
        InventoryState.categories.forEach(category => {
            const option = document.createElement('option');
            option.value = category;
            option.textContent = category;
            categoryFilter.appendChild(option);
        });
    }
    
    if (productCategory) {
        // Clear current options
        productCategory.innerHTML = '';
        
        // Add categories
        InventoryState.categories.forEach(category => {
            const option = document.createElement('option');
            option.value = category;
            option.textContent = category;
            productCategory.appendChild(option);
        });
        
        // Add option to create new category
        const newOption = document.createElement('option');
        newOption.value = 'new';
        newOption.textContent = '+ Add New Category';
        productCategory.appendChild(newOption);
        
        // Handle new category selection
        productCategory.addEventListener('change', function() {
            if (this.value === 'new') {
                const newCategory = prompt('Enter new category name:');
                if (newCategory && newCategory.trim()) {
                    // Add to category list if not exists
                    if (!InventoryState.categories.includes(newCategory.trim())) {
                        InventoryState.categories.push(newCategory.trim());
                        
                        // Add to dropdown
                        const option = document.createElement('option');
                        option.value = newCategory.trim();
                        option.textContent = newCategory.trim();
                        productCategory.insertBefore(option, productCategory.lastChild);
                        
                        // Select the new option
                        productCategory.value = newCategory.trim();
                        
                        // Also update filter dropdown
                        if (categoryFilter) {
                            const filterOption = document.createElement('option');
                            filterOption.value = newCategory.trim();
                            filterOption.textContent = newCategory.trim();
                            categoryFilter.appendChild(filterOption);
                        }
                    } else {
                        // Select existing category
                        productCategory.value = newCategory.trim();
                    }
                } else {
                    // If canceled or empty, revert to first option
                    productCategory.selectedIndex = 0;
                }
            }
        });
    }
}

/**
 * Render the product list
 */
function renderProductList() {
    const productListElem = document.getElementById('productList');
    if (!productListElem) return;
    
    // Clear current list
    productListElem.innerHTML = '';
    
    // Check if there are products
    if (InventoryState.products.length === 0) {
        productListElem.innerHTML = `
            <tr>
                <td colspan="6" class="text-center py-4">
                    <div class="alert alert-info mb-0">
                        No products found. Add your first product to get started.
                    </div>
                </td>
            </tr>
        `;
        return;
    }
    
    // Add products to list
    InventoryState.products.forEach(product => {
        // Determine stock status class
        let stockStatusClass = '';
        let stockStatusText = '';
        
        if (product.stock_quantity <= 0) {
            stockStatusClass = 'text-danger';
            stockStatusText = 'Out of Stock';
        } else if (product.stock_quantity <= 10) {
            stockStatusClass = 'text-warning';
            stockStatusText = 'Low Stock';
        } else {
            stockStatusClass = 'text-success';
            stockStatusText = 'In Stock';
        }
        
        const row = document.createElement('tr');
        row.innerHTML = `
            <td>${product.barcode}</td>
            <td>${product.name}</td>
            <td>${formatCurrency(product.price)}</td>
            <td>${product.category}</td>
            <td class="${stockStatusClass}">
                ${product.stock_quantity} 
                <small>(${stockStatusText})</small>
            </td>
            <td>
                <div class="btn-group btn-group-sm" role="group">
                    <button type="button" class="btn btn-outline-primary edit-product" data-id="${product.id}">
                        <i class="bi bi-pencil"></i>
                    </button>
                    <button type="button" class="btn btn-outline-danger delete-product" data-id="${product.id}">
                        <i class="bi bi-trash"></i>
                    </button>
                </div>
            </td>
        `;
        
        // Add event listeners
        row.querySelector('.edit-product').addEventListener('click', () => {
            editProduct(product.id);
        });
        
        row.querySelector('.delete-product').addEventListener('click', () => {
            deleteProduct(product.id);
        });
        
        productListElem.appendChild(row);
    });
    
    // Set up event listeners for buttons
    setupProductButtonListeners();
}

/**
 * Setup listeners for product action buttons
 */
function setupProductButtonListeners() {
    // Edit product buttons
    document.querySelectorAll('.edit-product').forEach(button => {
        button.addEventListener('click', function() {
            const productId = this.getAttribute('data-id');
            editProduct(productId);
        });
    });
    
    // Delete product buttons
    document.querySelectorAll('.delete-product').forEach(button => {
        button.addEventListener('click', function() {
            const productId = this.getAttribute('data-id');
            deleteProduct(productId);
        });
    });
}

/**
 * Render the stock movements list
 */
function renderStockMovements() {
    const stockListElem = document.getElementById('stockMovementsList');
    if (!stockListElem) return;
    
    // Clear current list
    stockListElem.innerHTML = '';
    
    // Check if there are stock movements
    if (InventoryState.stockMovements.length === 0) {
        stockListElem.innerHTML = `
            <tr>
                <td colspan="6" class="text-center py-4">
                    <div class="alert alert-info mb-0">
                        No stock movements found. Add stock adjustments to track inventory changes.
                    </div>
                </td>
            </tr>
        `;
        return;
    }
    
    // Add stock movements to list
    InventoryState.stockMovements.forEach(movement => {
        // Determine movement type class
        let typeClass = '';
        let typeIcon = '';
        
        if (movement.movement_type === 'in') {
            typeClass = 'text-success';
            typeIcon = 'bi-arrow-down-circle-fill';
        } else if (movement.movement_type === 'out') {
            typeClass = 'text-danger';
            typeIcon = 'bi-arrow-up-circle-fill';
        } else {
            typeClass = 'text-warning';
            typeIcon = 'bi-arrow-repeat';
        }
        
        const row = document.createElement('tr');
        row.innerHTML = `
            <td>${formatDate(movement.movement_date)}</td>
            <td>${movement.product_name}</td>
            <td class="${typeClass}">
                <i class="bi ${typeIcon}"></i>
                ${movement.movement_type.charAt(0).toUpperCase() + movement.movement_type.slice(1)}
            </td>
            <td class="${movement.quantity >= 0 ? 'text-success' : 'text-danger'}">
                ${movement.quantity}
            </td>
            <td>${movement.reference || '-'}</td>
            <td>${movement.notes || '-'}</td>
        `;
        
        stockListElem.appendChild(row);
    });
}

/**
 * Handle product form submission
 */
async function handleProductFormSubmit(event) {
    event.preventDefault();
    
    // Get form data
    const formData = new FormData(event.target);
    const productData = {
        barcode: formData.get('barcode'),
        name: formData.get('name'),
        description: formData.get('description'),
        price: parseFloat(formData.get('price')),
        cost_price: parseFloat(formData.get('costPrice')),
        category: formData.get('category'),
        stock_quantity: parseInt(formData.get('stockQuantity') || 0)
    };
    
    try {
        let response;
        const productId = formData.get('productId');
        
        if (productId) {
            // Update existing product
            response = await apiCall(`/inventory/products/${productId}`, {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(productData)
            });
            
            showToast('Success', 'Product updated successfully', 'success');
        } else {
            // Create new product
            response = await apiCall('/inventory/products', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(productData)
            });
            
            showToast('Success', 'Product added successfully', 'success');
        }
        
        // Close modal
        const productModal = bootstrap.Modal.getInstance(document.getElementById('productModal'));
        productModal.hide();
        
        // Refresh product list
        await loadProducts();
    } catch (error) {
        console.error('Error saving product:', error);
    }
}

/**
 * Handle stock movement form submission
 */
async function handleStockMovementSubmit(event) {
    event.preventDefault();
    
    // Get form data
    const formData = new FormData(event.target);
    const movementData = {
        product_id: parseInt(formData.get('productId')),
        quantity: parseInt(formData.get('quantity')),
        movement_type: formData.get('movementType'),
        reference: formData.get('reference'),
        notes: formData.get('notes')
    };
    
    try {
        // Create stock movement
        const response = await apiCall('/inventory/stock-movements', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(movementData)
        });
        
        showToast('Success', 'Stock movement recorded successfully', 'success');
        
        // Close modal
        const stockModal = bootstrap.Modal.getInstance(document.getElementById('stockMovementModal'));
        stockModal.hide();
        
        // Refresh data
        await Promise.all([
            loadProducts(),
            loadStockMovements()
        ]);
    } catch (error) {
        console.error('Error recording stock movement:', error);
    }
}

/**
 * Edit a product
 * @param {number} productId - Product ID to edit
 */
async function editProduct(productId) {
    try {
        // Find product in state
        const product = InventoryState.products.find(p => p.id == productId);
        if (!product) {
            // If not in state, fetch from API
            const response = await apiCall(`/inventory/products/${productId}`);
            product = response.product;
        }
        
        if (product) {
            // Populate form
            const form = document.getElementById('productForm');
            form.elements['productId'].value = product.id;
            form.elements['barcode'].value = product.barcode;
            form.elements['name'].value = product.name;
            form.elements['description'].value = product.description || '';
            form.elements['price'].value = product.price;
            form.elements['costPrice'].value = product.cost_price || 0;
            form.elements['category'].value = product.category;
            form.elements['stockQuantity'].value = product.stock_quantity;
            
            // Update form title
            document.getElementById('productModalLabel').textContent = 'Edit Product';
            
            // Show modal
            const productModal = new bootstrap.Modal(document.getElementById('productModal'));
            productModal.show();
        }
    } catch (error) {
        console.error('Error editing product:', error);
        showToast('Error', 'Failed to load product details', 'error');
    }
}

/**
 * Delete a product
 * @param {number} productId - Product ID to delete
 */
async function deleteProduct(productId) {
    // Confirm deletion
    if (!confirm('Are you sure you want to delete this product? This action cannot be undone.')) {
        return;
    }
    
    try {
        // Delete product
        await apiCall(`/inventory/products/${productId}`, {
            method: 'DELETE'
        });
        
        showToast('Success', 'Product deleted successfully', 'success');
        
        // Refresh product list
        await loadProducts();
    } catch (error) {
        console.error('Error deleting product:', error);
    }
}

/**
 * Reset the product form
 */
function resetProductForm() {
    const form = document.getElementById('productForm');
    if (form) {
        form.reset();
        form.elements['productId'].value = '';
        document.getElementById('productModalLabel').textContent = 'Add New Product';
    }
}

/**
 * Reset the stock movement form
 */
function resetStockMovementForm() {
    const form = document.getElementById('stockMovementForm');
    if (form) {
        form.reset();
        
        // Populate product dropdown
        const productSelect = document.getElementById('stockProductId');
        productSelect.innerHTML = '<option value="">Select Product</option>';
        
        InventoryState.products.forEach(product => {
            const option = document.createElement('option');
            option.value = product.id;
            option.textContent = `${product.name} (Stock: ${product.stock_quantity})`;
            productSelect.appendChild(option);
        });
    }
}

/**
 * Handle inventory search
 */
function handleInventorySearch(event) {
    const searchTerm = event.target.value.trim().toLowerCase();
    
    if (InventoryState.currentView === 'products') {
        // Filter products
        const filteredProducts = InventoryState.products.filter(product => {
            return (
                product.name.toLowerCase().includes(searchTerm) ||
                product.barcode.toLowerCase().includes(searchTerm) ||
                product.category.toLowerCase().includes(searchTerm) ||
                (product.description && product.description.toLowerCase().includes(searchTerm))
            );
        });
        
        // Update view with filtered products
        const tempProducts = InventoryState.products;
        InventoryState.products = filteredProducts;
        renderProductList();
        InventoryState.products = tempProducts;
    } else {
        // Filter stock movements
        const filteredMovements = InventoryState.stockMovements.filter(movement => {
            return (
                movement.product_name.toLowerCase().includes(searchTerm) ||
                movement.reference.toLowerCase().includes(searchTerm) ||
                (movement.notes && movement.notes.toLowerCase().includes(searchTerm))
            );
        });
        
        // Update view with filtered movements
        const tempMovements = InventoryState.stockMovements;
        InventoryState.stockMovements = filteredMovements;
        renderStockMovements();
        InventoryState.stockMovements = tempMovements;
    }
}

/**
 * Handle category filter change
 */
function handleCategoryFilter(event) {
    const category = event.target.value;
    
    if (category === 'all') {
        // Show all products
        renderProductList();
    } else {
        // Filter products by category
        const filteredProducts = InventoryState.products.filter(product => product.category === category);
        
        // Update view with filtered products
        const tempProducts = InventoryState.products;
        InventoryState.products = filteredProducts;
        renderProductList();
        InventoryState.products = tempProducts;
    }
}
